EditGravatar: Add tracking

v18 opened the pull request:
This PR adds tracking to the `EditGravatar` component, and to the new sidebar Gravatar image. Here are the events introduced with their descriptions:

Tracks:
- clicking the Gravatar button: `calypso_edit_gravatar_click` with the parameter userVerified
- receiving a file from the user (via FilePicker or DropZone): `calypso_edit_gravatar_file_receive`
- an error when getting an image from the user: `calypso_edit_gravatar_file_recieve_failure`
- starting the image upload: `calypso_edit_gravatar_upload_start`
- new Gravatar uploaded successfully: `calypso_edit_gravatar_upload_success`
- an error uploading the new Gravatar: `calypso_edit_gravatar_upload_failure`

statBump:
- errors when trying to change Gravatar using `calypso_gravatar_update_error`. Error types are:
  - `unsuccessful_http_response`
  - `bad_filetype`
  - `image_editor_error`
  - `no_bearer_token`

cc @gwwar, @Essayoh, and @travisw for a review of the new events

Essayoh commented:
Looks good to me, Val

gwwar commented on client/state/current-user/gravatar-status/actions.js:
```diff
@@ -34,20 +41,40 @@ export function uploadGravatar( file, bearerToken, email ) {
 					type: GRAVATAR_UPLOAD_REQUEST_FAILURE
 				} );
+				dispatch( recordTracksEvent( 'calypso_edit_gravatar_upload_failure' ) );
+				dispatch( bumpStat( 'calypso_gravatar_update_error', 'unsuccessful_http_response' ) );
```
Since we're dispatching an action here, `GRAVATAR_UPLOAD_REQUEST_FAILURE` we can attach analytics as metadata to that action.

```jsx
const gravatarUploadFailure = flowRight(
	withAnalytics( recordTracksEvent( 'calypso_edit_gravatar_upload_failure' ) ),
	withAnalytics( bumpStat( 'calypso_gravatar_update_error', 'unsuccessful_http_response' ) ),
	() => ( { type: 'GRAVATAR_UPLOAD_REQUEST_FAILURE' } )
)();
dispatch( gravatarUploadFailure );
```

see [client/state/analytics](https://github.com/Automattic/wp-calypso/blob/master/client/state/analytics/README.md) for more examples

gwwar replied:
@dmsnell the above example usage feels a bit awkward to me. Have we considered adding a function that accepts any number of analytics actions?

The signature might look something like:
```jsx
( action, ...analytics ) => ( {
```

dmsnell replied:
I didn't build one in but it'd be easy enough (question whether worth it)

```js
// inline
[
	recordTracksEvent( 'failure' ),
	bumpStat( 'update_error', 'blarg' ),
].reduce( withAnalytics, gravatarUploadFailure )

// with composeAnalytics
withAnalytics( composeAnalytics(
	recordTracksEvent( 'failure' ),
	bumpStat( 'upload_error', 'blarg' ),
), gravatarUploadFailure )
```

gwwar replied:
> please also note that as we move towards the use of the data layer and generalized action-based middleware these concerns get further separated. in fact, even now it may not be ideal to be tracking these events here instead of in the analytics middleware since it can watch for the gravatarUploadFailure action and send off the appropriate analytics.

Ah as another dispatch within the wpcom data handlers? That makes sense and should cover most cases where we call out to analytics. 

Note that we'll likely see a few odd analytics events if they're not related to any http requests. For example the `clickButton` event that tracks when we open the file picker.

v18 replied:
@dmsnell just to double-check your meaning here - you want to use the [analytics middleware  ](https://github.com/Automattic/wp-calypso/blob/master/client/state/analytics/middleware.js)in this case, similar to how we handle [notices](https://github.com/Automattic/wp-calypso/blob/master/client/state/notices/middleware.js)?

v18 replied:
@dmsnell and I talked about this, and since using the analytics middleware is a big undertaking, I kept the analytics actions in the same place. I updated them to use `withAnalyitcs` in b077952b34447c5e2cc57d7b3094bb09c02819c6

gwwar commented on client/state/current-user/gravatar-status/actions.js:
```diff
@@ -34,20 +41,40 @@ export function uploadGravatar( file, bearerToken, email ) {
 					dispatch( {
 						type: GRAVATAR_UPLOAD_REQUEST_SUCCESS
 					} );
+					dispatch( recordTracksEvent( 'calypso_edit_gravatar_upload_success' ) );
```
Similarly it may make sense to attach this to `GRAVATAR_UPLOAD_REQUEST_SUCCESS`. Usage is a bit shorter here since there's only one stat to bump.
```jsx
dispatch( withAnalytics( 
    recordTracksEvent( 'calypso_edit_gravatar_upload_success' ),
   { type: GRAVATAR_UPLOAD_REQUEST_SUCCESS }
) );
```

v18 replied:
Updated in b077952b34447c5e2cc57d7b3094bb09c02819c6

gwwar commented on client/state/current-user/gravatar-status/actions.js:
```diff
@@ -34,20 +41,40 @@ export function uploadGravatar( file, bearerToken, email ) {
+	};
+}
+
+export function clickButton( { isVerified } ) {
```
This feels like this handler should live in the related component.

v18 replied:
👍 Moved this and `receiveImage` to `<EditGravatar />` in 0777aad6e1c61ec2d600da56d0f99a9437aaff4a

gwwar commented on client/state/current-user/gravatar-status/actions.js:
```diff
@@ -34,20 +41,40 @@ export function uploadGravatar( file, bearerToken, email ) {
+export function clickButton( { isVerified } ) {
+	return dispatch => {
+		dispatch( recordTracksEvent( 'calypso_edit_gravatar_click', { userVerified: isVerified } ) );
+		dispatch( recordGoogleEvent( 'Me', 'Clicked on Edit Gravatar Button in Profile' ) );
```
Note that we have access to props in mapDispatchToProps:
```jsx
const mapDispatchToProps = ( dispatch, ownProps ) => {
    const isVerified = get( ownProps, 'user.isVerified' );
    //...
}; 
```

For a single stat bump the following is also an option:
```
const mapDispatchToProps = {
    uploadGravatarClick: partial( recordGoogleEvent, 'Me', 'Clicked on Edit Gravatar Button in Profile' );
};
```

gwwar replied:
`composeAnalytics` might be handy here as well.

v18 replied:
Updated in b077952b34447c5e2cc57d7b3094bb09c02819c6

v18 replied:
Didn't end up using `ownProps` since the `user` prop is coming from `mapStateToProps` and is not available in `ownProps`, so we'd to add the `mergeProps` argument. I opted for a parameter on the click button fn as before.

gwwar commented on client/state/current-user/gravatar-status/actions.js:
```diff
@@ -34,20 +41,40 @@ export function uploadGravatar( file, bearerToken, email ) {
+
+export function receiveImage() {
+	return dispatch => {
+		dispatch( recordTracksEvent( 'calypso_edit_gravatar_file_receive' ) );
```
For a single stat bump the following is also an option:
```jsx
const mapDispatchToProps = {
    uploadGravatarClick: partial( recordGoogleEvent, 'Me', 'Clicked on Edit Gravatar Button in Profile' );
};
```

gwwar commented:
I would recommend either providing more unique names for (`clickButton` and `receiveImage`) , or moving them to the related component.

dmsnell commented on client/blocks/edit-gravatar/test/index.jsx:
```diff
@@ -118,6 +118,7 @@ describe( 'EditGravatar', function() {
 					<EditGravatar
 						translate={ noop }
 						user={ user }
+						receiveImage={ noop }
```
is this something that could be a default prop? if so we wouldn't have to add it to every test

v18 replied:
These functions don't need a default value since they're added by `mapDispatchToProps`, and they only record analytics events. I hesitate to add default values only to make the tests neater, since in general we avoid changing app code for tests' sake in Calypso.

v18 commented:
@gwwar thanks! Updated the analytics code to use `withAnalytics` and moved some of it back to the component.

v18 commented:
thanks @gwwar & @dmsnell! I marked this as ready to merge - going to wait for my other Gravatar PRs before merging, since they're all pretty close now :)

v18 commented:
Squashed & rebased on `update/edit-gravatar-dropzone`

v18 commented:
Found and fixed couple of issues:
1. `mapDispatchToProps` was being used incorrectly, fixed in 25ab04cdfefb8df7054e66d616876dd7506deaba
2. Using `flowRight` with multiple `withAnalytics` calls resulted in an extra property (`group`) on the first analytics object (in `meta.analytics`). This is because the action used for the merge ([ref](https://github.com/Automattic/wp-calypso/blob/master/client/state/analytics/actions.js#L35)) on also has a meta property, which may had the extra prop. Fixed in 10588dc842185f6a8efc6ae937b83446aa9854ac by using `composeAnalytics` instead.

v18 commented:
New changes approved by @dmsnell ^ :)
